### ISSUE TITLE
[PROPOSAL] Write index on each build during development

### ISSUE THREAD

**Issue by elucid** [ember-cli/ember-cli-deploy#112]
I use ember-cli-deploy for a project that is tightly coupled with its API server. The API server injects some session state into the `<head>` which helps to bootstrap the app on pageload.

In development, I originally used proxy middleware to read the index page and allow the apps to be developed simultaneously. However, this was cumbersome and substantially different from what runs in production mode.

After watching a talk and learning about the `postBuild` hook for addons I wrote a little in-repo addon that writes the index to redis on each build in development so that the controller which reads from redis to serve the app can be used without any special case code in all environments.

I've just extracted this to a standalone addon at https://github.com/elucid/ember-cli-redis-proxy

If many ember-cli-deploy users closely develop their API and ember apps this might be useful to include in ember-cli-deploy itself. I mentioned this to @lukemelia at EmberConf and he suggested I open an issue here to discuss.

The addon above obviously only works for redis and doesn't make use of anything in ember-cli-deploy aside from using its config to connect to redis. If there is interest in including this in ember-cli-deploy I can make a PR which makes use of the index adapter configured for ember-cli-deploy, supports index stores other than redis, etc.

I haven't bothered doing any of those things because the simple addon works for my use case and without first gauging interest sending a PR would be quite presumptuous :) It possibly makes sense to just keep it as a separate addon.

**Comment by lukemelia** [comment]
I really like this idea. Maybe it could be controlled by of a `deployOnBuild` configuration option.

**Comment by elucid** [comment]
Yeah a flag makes sense--doing this automatically would not be desirable for certain index adapters.

**Comment by LevelbossMike** [comment]
Adding this to ember-deploy-redis behind a flag seems like a good idea to me. :+1:

**Comment by elucid** [comment]
Should it belong to that index adapter specifically? Granted the redis adapter is the only index adapter currently available where doing this makes sense but I could imagine other future adapters might want the same behaviour. I think it could be implemented using the interface all index adapters conform to.

**Comment by himynameisjonas** [comment]
@elucid awesome stuff!

**Comment by lukemelia** [comment]
I've implemented this on my project in 0.5.0 beta by defining a `devindex` deploy which only runs the `revision-key` and `redis` plugins, and adding an in-repo-addon with a `postInstall` hook that executes `ember deploy devindex --activate`. I'm open to making it even easier in 0.5.0, too, if people have ideas.

**Comment by himynameisjonas** [comment]
@lukemelia I tried to do what you did but failed on the in-repo-addon, would you be nice and share yours?

**Comment by elucid** [comment]
@lukemelia the solution you are describing is only run ad hoc, not after each change in the app, correct?

**Comment by himynameisjonas** [comment]
Ok, now I have something that may not be perfect but it works.
My in-repo-addon:
```js
var spawn = require('child_process').spawn;

module.exports = {
  name: 'redis-on-build',

  postBuild: function(result){
    var child = spawn('ember', ['deploy', 'development', result.directory]);

    child.stdout.on('data', function(data) {
      process.stdout.write(data.toString());
    });

    child.stderr.on('data', function(data) {
      process.stdout.write(data.toString());
    });
  },
};
```

And in my config/deploy.js:
```js
if (deployTarget === 'development') {
  var distDir = process.argv[process.argv.length - 1] // passed from in-repo-addon
  ENV.redis.distDir = distDir;
  ENV['json-config'].distDir = distDir.split('ember_app')[1];
  ENV.redis.url = process.env.REDIS_URL || 'redis://0.0.0.0:6379/';
  ENV.redis.filePattern = 'index.json';
  ENV.plugins = ['json-config', 'redis']; // only care about deploying index.json into redis in dev
}
```

But now, it spawns a new process to run the deploy command, is it possible to trigger that from the same process and avoid the extra time it takes to start the new process?

**Comment by lukemelia** [comment]
Checkout work-in-progress on #222

**Comment by lukemelia** [comment]
Fixed by #222
